include/zephyr/rtio/rtio.h

```diff
@@ -0,0 +1,418 @@
+};
+
+/**
+ * @brief A completion queue event
+ */
+struct rtio_cqe {
+	int32_t result; /**< Result from operation */
+	void *userdata; /**< Associated userdata with operation */
+};
+
+/**
+ * @brief Completion queue
+ *
+ * This is used for typifying the members of an RTIO queue pair
+ * but nothing more.
+ */
+struct rtio_cq {
+	struct rtio_spsc _spsc;
+	struct rtio_cqe buffer[];
+};
+
+struct rtio;
+
+struct rtio_executor_api {
+	int (*execute)(struct rtio *r, uint32_t wait_count);
+};
+
+/**
+ * @brief An executor does the work of executing the submissions.
+ *
+ * This could be a DMA controller backed executor, thread backed,
+ * or simple in place executor.
+ *
+ * A DMA executor might schedule all transfers with priorities
+ * and use hardware arbitration.
+ *
+ * A threaded executor might use a thread pool where each transfer
+ * chain is executed across the thread pool and the priority of the
+ * transfer is used as the thread priority.
+ *
+ * A simple in place exector might simply loop over and execute each
+ * transfer in the calling threads context. Priority is entirely
+ * derived from the calling thread then.
+ *
+ * An implementation of the executor must place this struct as
+ * its first member such that pointer aliasing works.
+ */
+struct rtio_executor {
+	const struct rtio_executor_api *api;
+};
+
+/**
+ * @brief An RTIO queue pair that both the kernel and application work with
+ *
+ * The kernel is the consumer of the submission queue, and producer of the completion queue.
+ * The application is the consumer of the completion queue and producer of the submission queue.
+ *
+ * Nothing is done until a call is performed to do the work (rtio_execute).
+ */
+struct rtio {
+
+	/*
+	 * An executor which does the job of working through the submission
+	 * queue.
+	 */
+	struct rtio_executor *executor;
+
+	/* Number of completions that were unable to be submitted with results
+	 * due to the cq spsc being full
+	 */
+	atomic_t xcqcnt;
+
+	/* Submission queue */
+	struct rtio_sq *sq;
+
+	/* Completion queue */
+	struct rtio_cq *cq;
+};
+
+/**
+ * @brief RTIO Submission Flags
+ * @defgroup rtio_submission_flags RTIO Submission Flags
+ * @ingroup rtio_api
+ * @{
+ */
+
+/**
+ * @brief Submit a request to be done while blocking the caller
+ *
+ * The submitted operation will be done synchronously and the call shall
+ * not return until the operation is complete. The default is a non-blocking
+ * request.
+ */
+#define RTIO_SUBMIT_SYNC BIT(0)
+
+/**
+ * @}
+ */
+
+
+/**
+ * @brief API that an RTIO IO device should implement
+ */
+struct rtio_iodev_api {
+	/**
+	 * @brief Submission function for a request to the iodev
+	 *
+	 * By default with no flags provided the function should
+	 * return as soon as possible. It is acceptable but not
+	 * ideal if the function must poll on hardware for completion
+	 * of the request.
+	 *
+	 * If the RTIO_SUBMIT_SYNC flag is set the function MUST
+	 * complete the request before returning.
```

Comment:
I keep seeing the same pattern that implies this flag only bottlenecks on this specific request and not any ASYNC requests that also preceeded this request. This could make the idea of ASYNC unusable in some instances.

Reply:
This flag is going to go away in the next iteration of this, which should address most of the concerns surrounding it I think. It was there initially to enable the simple executor loop I had created with a wait count that used a blocking call until the wait count was completed. The thinking was you are waiting on the count anyways so may as well block the calling thread until the requested wait count of submissions was complete.

Your commentary has me really thinking this is wrong, as really all operations should be asynchronous and the wait count done in a different method that allows the calling thread to sleep if desired (though polling is still possible)